Sources/GRPCCore/Documentation.docc/Articles/Migration-guide.md

```diff
@@ -0,0 +1,501 @@
+server configured to use the HTTP/2 transport:
+
+```swift
+let server = GRPCServer(
+  transport: .http2NIOPosix(
+    // Configure the host and port to listen on.
+    address: .ipv4(host: "127.0.0.1", port: 1234),
+    // Configure TLS here, if your're using it.
+    transportSecurity: .plaintext,
+    config: .defaults { config in
+      // Change any of the default config in here.
+    }
+  ),
+  // List your services here:
+  services: []
+)
+
+// Start the server.
+try await server.serve()
+```
+
+You can get the listening address using the `listeningAddress` property:
+
+```swift
+try await withThrowingDiscardingTaskGroup { group in
+  group.addTask { try await server.serve() }
+  if let address = try await server.listeningAddress {
+    print("Listening on \(address)")
+  }
+}
+```
+
+With any luck your app should build and your server should run. Yes, you guessed
+it, it's time to commit any changes you've made.
+
+## Client code migration
+
+> You can skip this section if you only needed to migrate services.
+
+Migrating client code is more difficult as you typically use client code
+throughout a wider part of your app. Our approach is to migrate from client
+calls first and then work upwards through your app to where the client is
+created.
+
+Start by finding a place within the target being migrated where a generated
+client is being used.
+
+Note that the generated client in 2.x is generic over a transport type, any
+types or functions using it will either need to choose a concrete type or
+also become generic. The most similar replacements to 1.x are:
+
+- `HTTP2ClientTransport.Posix`, and
+- `HTTP2ClientTransport.TransportServices`.
+
+Changing the type of the client will cause numerous build errors. To keep the
+number of errors manageable you'll migrate one function at a time. How this
+is done depends on whether the generated client is passed in to the function
+or stored on a property.
+
+If the function is passed a generated client then duplicate it, changing the
+signature to use a 2.x generated client. The new client is
+named `{Service}.Client` where `{Service}` is the namespaced name of your
+service (if your service is named `Baz` and declared in the `foo.bar`
+Protocol Buffers package then this would be `Foo_Bar_Baz.Client`).
+Change the body of the function using the 1.x client to just `fatalError()`.
+Later you'll remove this function altogether.
+
+If the generated client is a stored type then add a new computed property
+returning an instance of it, the body can just call `fatalError()` for now:
+
+```swift
+var client: Foo_Bar_Baz.Client {
+  fatalError("TODO")
+}
+```
+
+If a generated client is passed into the function then duplicate the function
+and replace the body of the new version with a `fatalError()`. You can also mark
+it as deprecated to help you find usages of the function. You'll now have two
+versions of the same function.
```

Comment:
Feels like this paragraph and the one above beginning with "If the function is passed a generated client then duplicate it" should be merged.